### PR TITLE
Add WooCommerce API Create Product

### PR DESCRIPTION
**Pull request by coderkevin**
This is the supporting code needed to trigger an API call and handle it within the action and reducer code.

Update: The previous problem with the WPCOM API has been solved!

This takes care of the code up to and including sending off the API Create request.
Some error handling and notices are needed to be added next.

### CODE REVIEW

File: client/extensions/woocommerce/app/products/product-header.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import Card from 'components/card';
+import Button from 'components/button';
+import StickyPanel from 'components/sticky-panel';
+
+const ProductHeader = ( { onTrash, onDuplicate, onSave, translate } ) => {
+	const trashButton = onTrash &&
+		<Button borderless onClick={ onTrash }><Gridicon icon="trash" /></Button>;
+
+	const duplicateButton = onDuplicate &&
+		<Button onClick={ onDuplicate }>{ translate( 'Duplicate' ) }</Button>;
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/product-header.js]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Skipping duplicate' )` *ES Score: 13*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/state/wc-api/products/actions.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+		dispatch( createAction );
+
+		// Filter out any id we might have.
+		const { id, ...productData } = product;
+
+		const jpProps = { path: `/jetpack-blogs/${ siteId }/rest-api/` };
+		const httpProps = {
+			path: '/wc/v2/products',
+			body: productData,
```
Comment (oskosk) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: This should probably be stringified:
```
body: JSON.stringify( productData )
```

Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: 👍

File: client/extensions/woocommerce/style.scss [hunk 3]
```diff
@@ -26,6 +26,20 @@
 		border: 0;
 	}
 
+	.products__header {
```
Comment (belcherj) [inline client/extensions/woocommerce/style.scss]: This is general comment and not on this PR.  Will this css file not get massive?  Is there a mechanism to break this up into the components?

File: client/extensions/woocommerce/state/reducer.js [hunk 4]
```diff
@@ -7,7 +7,9 @@ import { combineReducers } from 'redux';
  * Internal dependencies
  */
 import ui from './ui/reducer';
+import wcApi from './wc-api/reducer';
 
 export default combineReducers( {
-	ui
+	ui,
+	wcApi,
 } );
```
Comment (oskosk) [inline client/extensions/woocommerce/state/reducer.js]: yes, it is allowed

File: client/extensions/woocommerce/app/products/product-header.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+	const saveButton = onSave &&
+		<Button primary onClick={ onSave }>{ translate( 'Save' ) }</Button>;
+
+	return (
+		<StickyPanel>
+			<Card className="products__header">
+				<span>Breadcrumbs > go > here</span>
```
Comment (coderkevin) [inline client/extensions/woocommerce/app/products/product-header.js]: Fair enough. Added.

File: client/extensions/woocommerce/state/reducer.js (hunk 4 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/reducer.js]: Yeah, I guess I missed it the first time, but this corrects it.

File: client/extensions/woocommerce/app/products/product-header.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Internal dependencies
+ */
+import Card from 'components/card';
+import Button from 'components/button';
+import StickyPanel from 'components/sticky-panel';
+
+const ProductHeader = ( { onTrash, onDuplicate, onSave, translate } ) => {
+	const trashButton = onTrash &&
```
Comment (coderkevin) [inline client/extensions/woocommerce/app/products/product-header.js]: The functions are optional, and their presence indicates the button should be present.

File: client/extensions/woocommerce/state/wc-api/products/actions.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+			type: WOOCOMMERCE_API_CREATE_PRODUCT,
+			payload: { siteId, product },
+		};
+
+		dispatch( createAction );
+
+		// Filter out any id we might have.
+		const { id, ...productData } = product;
+
+		const jpProps = { path: `/jetpack-blogs/${ siteId }/rest-api/` };
```
Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: Okay, that's fair.

File: client/extensions/woocommerce/style.scss (hunk 3 above)
Comment (coderkevin) [inline client/extensions/woocommerce/style.scss]: I was thinking the same thing. I think we'll have to split this out.

File: client/extensions/woocommerce/state/action-types.js [hunk 8]
```diff
@@ -1,3 +1,7 @@
+export const WOOCOMMERCE_API_SET_ERROR = 'WOOCOMMERCE_API_SET_ERROR';
+export const WOOCOMMERCE_API_CLEAR_ERROR = 'WOOCOMMERCE_API_CLEAR_ERROR';
+export const WOOCOMMERCE_API_CREATE_PRODUCT = 'WOOCOMMERCE_API_CREATE_PRODUCT';
```
Comment (v18) [inline client/extensions/woocommerce/state/action-types.js]: A small note: `WOOCOMMERCE_API_CREATE_PRODUCT_REQUEST` would conform better with the latest Calypso naming conventions.

File: client/extensions/woocommerce/state/wc-api/products/actions.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		}
+
+		const jetpackProps = { path: `/jetpack-blogs/${ siteId }/rest-api/` };
+		const httpProps = {
+			path: '/wc/v2/products',
+			body: JSON.stringify( productData ),
+			json: true,
+		};
+
+		return wp.req.post( jetpackProps, httpProps )
```
Comment (v18) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: From what I know, Calypso is moving to making network requests using the [data layer](https://github.com/Automattic/wp-calypso/tree/master/client/state/data-layer). This is so that we can offer offline support & better managing multiple requests.

I just did a migration to the data layer for another component, so would be happy to help.

cc @dmsnell for a more nuanced opinion.

Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: I just recently added a data layer middleware for extensions. I plan on using it after some changes with the data layer go through, so the thunks are just temporary. But I did just add a couple `// TODO:` comments to reflect that. Thanks!

File: client/extensions/woocommerce/state/action-types.js (hunk 8 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/action-types.js]: Hmm, that might be a good thing to do. But if we decide to do that, it should probably be a separate PR just to rename actions.

**Comment by coderkevin** [comment]
Rebased and added a couple `// TODO` comments.

**Comment by coderkevin** [comment]
@justinshreve, @allendav, @belcherj, @DanReyLop, I think this is ready for review/merge at this point. There's still more to be done, but I'll be opening a separate PR for that.

File: client/extensions/woocommerce/state/wc-api/products/actions.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		};
+
+		// TODO: Modify this to use the extensions data layer.
+		return wp.req.post( jetpackProps, httpProps )
+			.then( ( { data } ) => {
+				dispatch( createProductSuccess( siteId, data ) );
+			} )
+			.catch( err => {
+				dispatch( error( siteId, createAction, err ) );
+			} );
```
Comment (dmsnell) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: pardon my coming into this at the last moment here but did you consider using the HTTP system in the data layer?

Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: Certainly. But when I started this process, the data layer for extensions hadn't yet been merged. I figured it was best to start simple and upgrade to using the data layer for such things afterwards. I've added #14280 to track this.

Comment (dmsnell) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: :+1:

**Comment by belcherj** [comment]
@coderkevin where does this stand?  Tests failed on CircleCI, once they pass I'll review again.

**Comment by coderkevin** [comment]
Linting issue fixed and rebased.

**Comment by coderkevin** [comment]
@belcherj @allendav This is ready for review

**Comment by coderkevin** [comment]
Rebased

File: client/extensions/woocommerce/app/products/product-create.js [hunk 11]
```diff
@@ -61,7 +62,8 @@ class ProductCreate extends React.Component {
 	}
 
 	onSave = () => {
-		// TODO: Add action dispatch to save this product.
+		const { siteId, product } = this.props;
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-create.js]: does this component need to know site id? Shouldn't create product be able to get siteId or at least have a default.  @DanReyLop and I discussed here: https://github.com/Automattic/wp-calypso/pull/14438#discussion_r118713391

Comment (belcherj) [inline client/extensions/woocommerce/app/products/product-create.js]: Follow up PR?

File: client/extensions/woocommerce/state/wc-api/products/actions.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+		if ( typeof id === 'number' ) {
+			dispatch( error( siteId, createAction, {
+				message: 'Attempting to create a product which already has a valid id.',
+				product,
+			} ) );
+			return;
+		}
+
+		const jetpackProps = { path: `/jetpack-blogs/${ siteId }/rest-api/` };
```
Comment (belcherj) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: trailing comma?

File: client/extensions/woocommerce/app/products/product-create.js (hunk 11 above)
Comment (coderkevin) [inline client/extensions/woocommerce/app/products/product-create.js]: Yeah, I'll definitely be looking to change the way siteId is tracked for the edit state.

File: client/extensions/woocommerce/state/wc-api/products/actions.js (hunk 12 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: I don't usually do those on one-liners. Hmm.

Comment (justinshreve) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: Yeah I think it's okay to leave off for 1 liners like this.

Comment (justinshreve) [inline client/extensions/woocommerce/state/wc-api/products/actions.js]: Yeah I think it's okay to leave off for 1 liners like this.
